ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/FutureAssertions.scala

```diff
@@ -34,13 +40,121 @@ final class FutureAssertions[T](future: Future[T]) {
       case Success(value) => Failure(new ExpectedFailureException(context, value))
       case Failure(other) => Failure(other)
     }
+
 }
 
 object FutureAssertions {
 
+  private val logger = ContextualizedLogger.get(getClass)
+
+  /** Runs the test case after the specified delay
+    */
+  def assertAfter[V](
+      delay: FiniteDuration
+  )(test: => Future[V]): Future[V] =
+    Delayed.Future.by(delay)(test)
+
+  /** Run the test every [[retryDelay]] up to [[maxRetryDuration]].
+    * The assertion will succeed if any of the test case runs are successful.
+    * The assertion will fail if no test case runs are successful and the [[maxRetryDuration]] is exceeded.
+    * The test case will run up to [[ceil(maxRetryDuration / retryDelay)]] times
+    */
+  def succeedsEventually[V](
+      retryDelay: FiniteDuration = 100.millis,
+      maxRetryDuration: FiniteDuration,
+      description: String,
+  )(
+      test: => Future[V]
+  )(implicit ec: ExecutionContext, loggingContext: LoggingContext): Future[V] = {
+    def internalSucceedsEventually(remainingDuration: FiniteDuration): Future[V] = {
+      val nextRetryRemainingDuration = remainingDuration - retryDelay
+      if (nextRetryRemainingDuration < Duration.Zero) test.andThen { case Failure(exception) =>
+        logger.error(
+          s"Assertion never succeeded after $maxRetryDuration with a delay of $retryDelay. Description: $description",
+          exception,
+        )
+      }
+      else
+        assertAfter(retryDelay)(test).recoverWith { case NonFatal(ex) =>
+          logger.debug(
+            s"Failed assertion: $description. Running again with new max duration $nextRetryRemainingDuration",
+            ex,
+          )
+          internalSucceedsEventually(nextRetryRemainingDuration)
+        }
+    }
+
+    internalSucceedsEventually(maxRetryDuration)
+  }
+
+  /** Run the test every [[rerunDelay]] for a duration of [[succeedDuration]] or until the current time exceeds [[succeedDeadline]].
+    * The assertion will succeed if all of the test case runs are successful and [[succeedDuration]] is exceeded or [[succeedDeadline]] is exceeded.
+    * The assertion will fail if any test case runs fail.
+    */
+  def succeedsUntil[V](
+      rerunDelay: FiniteDuration = 100.millis,
+      succeedDuration: FiniteDuration,
+      succeedDeadline: Option[Instant] = None,
```

Comment:
```suggestion
      duration: FiniteDuration,
      until: Option[Instant] = None,
```

Also, if you don't need to pass both, consider an `Either`.